### ISSUE TITLE
Racket fails to build if coreutils are not in /bin

### ISSUE THREAD

**Issue by ralsei** [racket/racket#3707]
**What version of Racket are you using?**
8.0.0.10 [cs] (git)

**What program did you run?**
When building Racket (with just `make`) on a system without cp in `/bin` (or a number of coreutils), the build fails.

**What should have happened?**
Racket probably should have compiled.

**If you got an error message, please include it here.**
```
make[9]: Entering directory '/home/hazel/tmp/racket/racket/src/build/cs/c'
cd ChezScheme && "../../../../cs/c/../../ChezScheme"/configure --pb --machine=ta6le --disable-x11 --disable-curses --disable-auto-flags  CC="clang" CFLAGS="-g -O2  -DELF_FIND_BOOT_SECTION  -pthread" LDFLAGS="-pthread" LIBS="-ldl -lm  -lrt" AR="ar" ARFLAGS="rc" RANLIB=":" WINDRES="windres"
Configuring for pb
../../../../cs/c/../../ChezScheme/workarea: line 281: /bin/cp: No such file or directory
../../../../cs/c/../../ChezScheme/workarea: line 287: /bin/cp: No such file or directory
make[9]: Leaving directory '/home/hazel/tmp/racket/racket/src/build/cs/c'
cd ChezScheme && make reset
make[9]: Entering directory '/home/hazel/tmp/racket/racket/src/build/cs/c/ChezScheme'
(cd pb && make reset)
(cd s && make reset)
make[11]: *** No rule to make target 'reset'.  Stop.
make[10]: *** [Makefile:50: reset] Error 2
```

**Please include any other relevant details**
NixOS solves this issue by patching various files to update the hardcoded paths: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/racket/default.nix#L74

BC had a similar issue, but only with `uname`: https://github.com/NixOS/nixpkgs/blob/nixos-20.09/pkgs/development/interpreters/racket/default.nix#L73

**Comment by LiberalArtist** [comment]
I discovered this issue while working on a [updating the GNU Guix package](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=46865) to Racket 8.0 (CS). I am working on a patch.

The part I'm least clear about is `uname`, which @ralsei points out applies both to BC and to CS. It seems to be used at run-time to answer `(system-type 'machine)`. See:

- CS
  - https://github.com/racket/racket/blob/46a191df038244199f83f2bb1235beb8ca137f9a/racket/src/io/machine/main.rkt
  - https://github.com/racket/racket/blob/4e05a0d058a8382dd3802b4c238a3ee8a9411493/racket/src/cs/rumble/system.ss#L124
- BC
  - https://github.com/racket/racket/blob/421dce228d5a98d89db2bddb27cd32211cdb1ae3/racket/src/bc/src/string.c#L2407-L2420
  - https://github.com/racket/racket/blob/421dce228d5a98d89db2bddb27cd32211cdb1ae3/racket/src/bc/src/string.c#L5527-L5535
  - https://github.com/racket/racket/blob/421dce228d5a98d89db2bddb27cd32211cdb1ae3/racket/src/bc/src/string.c#L5581

Would it be possible instead to use `uname(3)` from `libc` directly, perhaps from `rktio`, so it could be shared between CS and BC? According to my local `man` page, `uname(3)` was in POSIX.1 (1988), whereas `uname(1)` was added in POSIX.2 (1992), so I think it would be available everywhere.

Aside from supporting environments like Nix and Guix, using the C version would also avoid starting a subprocess and potentially trying multiple paths, so it seems like it would be an improvement for everyone.

**Comment by mflatt** [comment]
@LiberalArtist Will do.

**Comment by LiberalArtist** [comment]
I'm hopeful that https://github.com/racket/racket/pull/3710 will close this issue, but I plan to investigate further. In particular, it looks like Nix patches uses of `/usr/bin/uname` in [racket/src/lt/configure](https://github.com/racket/racket/blob/e615294e7866e145a511255cf93814df9406289d/racket/src/lt/configure). I hadn't done that for Guix and hadn't had problems, but I'm not clear whether I was actually running that script for a CS-only build.

**Comment by pmatos** [comment]
@LiberalArtist Is this problem with NixOS/Guix something I can reproduce this in a container? I have no experience with NixOS/Guix so I would need some pointers.

**Comment by LiberalArtist** [comment]
It will be!

Right now, Nix and Guix both apply the patches they need, so there isn't a "problem". When I was updating the package definition for Guix to 8.0, because of the change to make CS the default, there were a bunch of new patches needed, which prompted me to notice that a very large portion of these patches improve portability even outside of the Nix/Guix context and should be fixed in Racket (and then upstream Chez). For example, the [GNU Autoconf manual](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.70/html_node/The-Make-Macro-SHELL.html) gives other examples of obscure systems where `/bin/sh` is missing or is not a POSIX-compliant shell.

I'm not an expert in either, but Nix and Guix have an emphasis on reproducibility, bootstrapability, and explicit dependencies, so it seems like they could have value as a CI exercise for Racket. (I've used Nix a bit as an end user, and I've started exploring Racket packaging as a way to learn more about Guix, since I've heard from @cwebber and mailing-list discussions that Racket on Guix needs some love.)

I'm not sure which tools will eventually be most useful, but here are instructions on [installing Guix in a VM](https://guix.gnu.org/manual/en/html_node/Installing-Guix-in-a-VM.html#Installing-Guix-in-a-VM), using [`guix system`](https://guix.gnu.org/manual/en/html_node/Invoking-guix-system.html) to turn a system definition into a Docker image or various types of VM/disk image (`guix system image`/`guix system docker-image`) or running it in a Linux container (`guix system container`, and [packing packages+dependencies](https://guix.gnu.org/manual/en/html_node/Invoking-guix-pack.html#Invoking-guix-pack) into a Docker image or tarball.

I'm currently adapting my Guix package definition to work with a git checkout (rather than a source distribution) and exercise more of the Racket bootstrapping process, for which I'm looking at some of your CI work. I'll put that somewhere visible shortly (hopefully).

**Comment by ralsei** [comment]
@pmatos I encountered this on NixOS while building Racket from git with `make` in the environment `nix-shell '<nixpkgs>' -A racket`, if that's useful. After running some `sed` commands to fix the problematic files, it works fine -- but the point is that you shouldn't have to patch the Makefile (and some others).

**Comment by LiberalArtist** [comment]
Well, for those not familiar with Nix/Guix, there are some system-specific patches needed to make a package (not necessarily to build something in an interactive environment), like rewriting:
```sh
#!/bin/sh
``` 
to something like:
```
#!/gnu/store/pwcp239kjf7lnj5i4lkdzcfcxwcfyk72-bash-minimal-5.0.16/bin/sh
```
Those sorts of patches are well known to the relevant people and applying them is highly automated.

The issue is that Racket (and more so Chez Scheme) has, essentially, some portability bugs, e.g. (in Makefiles):
- writing `/bin/sh` rather than `$(SHELL)`;
- writing `cc` rather than `$(CC)`; and
- writing absolute paths for utilities that the GNU Make manual says can be [used directly](https://www.gnu.org/software/make/manual/html_node/Utilities-in-Makefiles.html#Utilities-in-Makefiles) with no need for [command variables](https://www.gnu.org/software/make/manual/html_node/Command-Variables.html#Command-Variables).

Nix and Guix expose such issues by using especially restricted build environments, and packagers for those systems maintained ad-hoc patches. But fixing these issues is a general improvement, so #3710 upstreamed patches for the issues I've discovered so far. (And the relevant changes should be further upstreamed to Chez, once I'm confident there aren't any more.)

**Comment by LiberalArtist** [comment]
@pmatos I've put my work in progress at bootstrapping Racket on Guix at <https://gitlab.com/philip1/guix-racket-experiment>, and my patch to the existing Guix Racket package is at <https://issues.guix.gnu.org/46865>.

**Comment by LiberalArtist** [comment]
I have been looking at `racket/src/lt/`, which `racket/src/README.txt` says contains “`libtool` and `configure` support”. If I've got this right, it contains:

  - `install-sh`: Seems to be a bundled implementation of the generic `install` command, like the one from GNU Coreutils or [this one](https://opensource.apple.com/source/gcc/gcc-5482/install-sh.auto.html).
 
  - `ltmain.sh`: The actual implementation of Libtool. Instead, you can opt to bring your own by passing `--enable-lt` to `racket/src/configure`.

  - `aclocal.m4`, `config.guess`, `config.sub`, `configure`, & `configure.in`: As best as I can tell, these seem to configure `ltmain.sh`, but there are some problems.

Guix would want to build with its bootstrapped coreutils, and I'd guess Nix would too, especially since [they've been patching](https://github.com/NixOS/nixpkgs/blob/4d0ee90c6e253d40920f8dae5edb717a7d6f151d/pkgs/development/interpreters/racket/default.nix#L75-L77) a use of `/usr/bin/uname` in `racket/src/lt/configure`. My first approach was to use `--enable-lt` and delete the `lt` directory. This exposed two issues with `racket/src/cfg-bc`:

1. It looks in `racket/src/lt/` for `install-sh`, `install.sh`, or `shtool` and fails if it can't find any of them: https://github.com/racket/racket/blob/46a191df038244199f83f2bb1235beb8ca137f9a/racket/src/cfg-bc#L1748-L1766
I was able to work around this by making `racket/src/lt/install-sh` a symlink to `install`, which worked fine. But it would be nicer if there were an option like `--enable-lt`, and it might be even nicer yet if `cfg-bc` could try `command -v install` to discover that there is one in its PATH, even if just as a fourth choice.
2. Immediately after that, it proceeds to “make sure we can run config.sub” and fails otherwise: https://github.com/racket/racket/blob/46a191df038244199f83f2bb1235beb8ca137f9a/racket/src/cfg-bc#L1768-L1825
Here I just don't understand what's going on. If `config.sub` was for configuring `ltmain.sh`, then presumably it shouldn't be relevant when `--enable-lt` is in effect. But maybe it's doing some other important checking that I just don't understand: I don't know much about autotools. In the mean time, I can just leave it there while deleting all of it's siblings, and the build works.

I guess these are probably questions for @mflatt.

**Comment by mflatt** [comment]
The files in `lt` include ones that are needed for `configure`, not (just) `libtool`. So, `lt` may be a misleading name for the directory, but it can't be simply deleted. The details about how files there are used are all part of `autoconf`, as opposed to something specific to Racket. 

Overall, I'm not following why you're trying to delete the files, instead of just setting `--enable-lt`.

**Comment by samth** [comment]
One question more broadly about some of the things @LiberalArtist mentions is whether we need libtool at all for building Racket, and if not whether we can make autoconf not including anything about it.

**Comment by mflatt** [comment]
libtool is only used when `--enable-shared` is specified, and only BC supports `--enable-shared`. (The CS configure script ignores/cancels `--enable-shared`.) I don't think `--enable-shared` is helpful for Racket BC, but people have insisted on using it, and probably BC should continue to support it in the way that it has.

**Comment by samth** [comment]
I think keeping `--enable-shared` is the right thing, despite its lack of value, but I'm wondering if we can do that without libtool. I'll investigate and see if there's something that makes sense to try out.

**Comment by ralsei** [comment]
Compiling Racket directly from the repo on NixOS with just `nix-shell -A '<nixpkgs>' racket` (read: installing dependencies) and then `make` works fine now from the latest commit.

**Comment by LiberalArtist** [comment]
A Guix user [reported an issue with fonts](https://issues.guix.gnu.org/47204): I haven't been able to reproduce it yet, but it seems unlikely to be related. It's been a while since I've tried to debug such issues with fontconfig and free desktop environments, though. @ralsei, I'd be interested if you've heard or seen anything from the Nix community.

Re @mflatt:
> The files in `lt` include ones that are needed for `configure`, not (just) `libtool`. So, `lt` may be a misleading name for the directory, but it can't be simply deleted. The details about how files there are used are all part of `autoconf`, as opposed to something specific to Racket.
> 
> Overall, I'm not following why you're trying to delete the files, instead of just setting `--enable-lt`.

I was probably being a bit overzealous in my interpretation of [Guix packaging guideline](https://guix.gnu.org/manual/en/html_node/Submitting-Patches.html) to "make sure the package does not use bundled copies of software already available as separate packages." I only have a fairly general understanding of the reasoning behind the guideline, but it would presumably reduce (or at least not increase) time and space for building, given the Guix/Nix "pure" dependency model, and I believe it's also important to the ["grafts" mechanism](https://guix.gnu.org/manual/en/html_node/Security-Updates.html#Security-Updates) for urgent security patches.

But, unless I hear otherwise from Guix people, it doesn't seem like we really need to worry about that here. I've just confirmed that CS, 3M, and CGC can all be build on Guix without changing anything under `racket/src/lt/`. (The builds also seem to work if I delete `racket/src/lt/ltmain.sh` and `racket/src/lt/install-sh` only and symlink `racket/src/lt/install-sh` to Guix's `install`—but I think I will stop doing that, anyway.)

My one remaining concern is that [here, Nix](https://github.com/NixOS/nixpkgs/blob/4d0ee90c6e253d40920f8dae5edb717a7d6f151d/pkgs/development/interpreters/racket/default.nix#L75-L77) is patching `src/lt/configure` and `src/cs/c/configure` to rewrite the path to `/usr/bin/uname`. I've not been doing that, and I haven't had any problems, but I don't know if that's because the patches aren't needed (good), because I haven't tested the right build mode (probably bad), or because something is defaulting to "unknown" (maybe ok).

**Comment by LiberalArtist** [comment]
Re @pmatos:
> @LiberalArtist Is this problem with NixOS/Guix something I can reproduce this in a container? I have no experience with NixOS/Guix so I would need some pointers.

I may have misunderstood your question. If you just want to set up Guix in a container, you can use it as a userspace package manager: it is now [in Debian Testing](https://packages.debian.org/bullseye/guix), or you can install it on top of most distros using [these instructions](https://guix.gnu.org/manual/en/html_node/Installation.html#Installation). Alternatively, there are [more instructions](https://guix.gnu.org/manual/en/html_node/System-Installation.html) for installing Guix as a standalone GNU/Linux system (or GNU/Hurd :) ), or there's an official pre-built QCOW2 image distributed here: https://ftp.gnu.org/gnu/guix/guix-system-vm-image-1.2.0.x86_64-linux.xz

**Comment by LiberalArtist** [comment]
@mflatt I was also going to make sure you saw that, when I proposed the changes from https://github.com/racket/racket/commit/fdb2ad7892933aea1eedd5b199432b988a7df972 upstream in https://github.com/cisco/ChezScheme/pull/555, @akeep suggested a few adjustments: mostly `rm -f` rather than `rm`, plus one fix for a format string. There's an un-squashed commit at https://github.com/LiberalArtist/ChezScheme/commit/6783e79f80071b8824ba846beef04b390fd287eb with just the changes in https://github.com/cisco/ChezScheme/commit/8f4633ce24ac6425b2ab13cc78026b1c9bb5361e that weren't in https://github.com/racket/racket/commit/fdb2ad7892933aea1eedd5b199432b988a7df972. (The second commit, https://github.com/cisco/ChezScheme/commit/63f40f61beef2623d3b69140e9a110d7ca67f278, is entirely new.) None of this is causing any problems for me; I just hope it doesn't cause any confusion when maintaining the fork.

**Comment by LiberalArtist** [comment]
Re @samth:
> I think keeping `--enable-shared` is the right thing, despite its lack of value, but I'm wondering if we can do that without libtool. I'll investigate and see if there's something that makes sense to try out.

I saw a recent [blog post from Andy Wingo](https://wingolog.org/archives/2021/04/08/sign-of-the-times) about ending Guile's use of `libltdl` and his hope to stop using libtool altogether:
> In summary, this is a case where the benefits of an abstraction layer decline over time. For a few years now, libltdl hasn't been paying for itself. Libtool is dead, for all intents and purposes (last release in 2015); best to make plans to migrate away, somehow.
> …
> Looking forward, I think it's probably time to replace Guile's use of libtool to create its libraries and executables. I loathe the fact that libtool puts shell scripts in the place of executables in build directories and stashes the actual executables elsewhere -- like, visceral revulsion. There is no need for that nowadays. Not sure what to replace it with, nor on what timeline.

This is out of my depth (and perhaps getting off-topic for this issue), but, in particular, Andy's remark that:
> Worse, as ltdl would search through the path for candidates, it would only report the _last_ error it saw from the underlying `dlopen` interface. It was almost always the case that if `A` and `B` were in the search path, and `A/foo.so` failed to load because of a missing dependency, the [error you would get as a user would instead be "file not found"](https://lists.gnu.org/archive/html/bug-libtool/2008-10/msg00017.html), because ltdl swallowed the first error and kept trucking to try to load `B/foo.so` which didn't exist.

recalled to me this caveat from the [docs for `ffi-lib`](https://docs.racket-lang.org/foreign/Loading_Foreign_Libraries.html#(def._((lib._ffi%2Funsafe..rkt)._ffi-lib))):

> A library file may exist but fail to load for some reason; the eventual error message will unfortunately name the fallback from the second or third bullet, since some operating systems offer no way to determine why a given library path failed.

I wonder if the simplifications Andy outlines in the library-loading landscape over the last several years might mean that it's possible to give a better error here.

**Comment by samth** [comment]
1. I think removing the use of libtool entirely would be reasonable but would be a nontrivial amount of hacking on the configure scripts, for limited benefit since they work now and will become less significant as Racket BC does  
2. I don't think we use libltdl dynamically for those situations, and thus I don't think things have gotten better on the error message front.

**Comment by LiberalArtist** [comment]
All of the patches Guix needed for Racket's build scripts can be removed as of Racket 8.1 (see http://issues.guix.gnu.org/48308), so this issue seems to be fixed.

**Comment by LiberalArtist** [comment]
Hmm, I will take another look later. @ralsei, was this the case in a fresh Git checkout, when re-building, or both? Maybe what I've been trying has been falling into the

https://github.com/racket/racket/blob/e0063a9495d1a1f657869ae281240e9985b7d055/racket/src/cs/c/check_boot.sh#L14

and

https://github.com/racket/racket/blob/e0063a9495d1a1f657869ae281240e9985b7d055/racket/src/cs/c/reset_boot.sh#L13

cases.

**Comment by ralsei** [comment]
Ah -- yeah, this wasn't a fresh checkout. I can try that.

**Comment by LiberalArtist** [comment]
That makes sense. It would be nice to make this work, but I'm not sure what the portable solution would is. Unlike the other commands I replaced, `find` isn't on [this list of utilities](https://www.gnu.org/software/make/manual/html_node/Utilities-in-Makefiles.html#Utilities-in-Makefiles) that the Make manual says are safe to use directly. On the other hand, [the Autoconf manual lists `find` here](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.70/html_node/Limitations-of-Usual-Tools.html), and the example seems to show using it directly as `find`, rather than `$(FIND)` or `/usr/bin/find`. I will say that `sh` is not one of my favorite programming lanugages …

**Comment by LiberalArtist** [comment]
We've been building Racket without these problems in Guix for some time now, and the offending use of `/usr/bin/find` for non-clean builds that @ralsei identified in https://github.com/racket/racket/issues/3707#issuecomment-835859551 was probably fixed in 43471471133be70c941d1a500f5b9f91c5d107dd and is in any case obsolete since the arrival of Zuo, so I think this issue can be closed.
